starboard/shared/libiamf/iamf_audio_decoder.cc

```diff
@@ -0,0 +1,382 @@
+      sample_info.discarded_duration_from_back);
+
+  decoded_audios_.push(decoded_audio);
+
+  output_cb_();
+
+  return true;
+}
+
+void IamfAudioDecoder::WriteEndOfStream() {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(output_cb_);
+
+  stream_ended_ = true;
+  if (!pending_audio_buffers_.empty()) {
+    return;
+  }
+
+  // Put EOS into the queue.
+  decoded_audios_.push(new DecodedAudio);
+
+  Schedule(output_cb_);
+}
+
+bool IamfAudioDecoder::InitializeCodec() {
+  SB_DCHECK(is_valid());
+  SB_DCHECK(!decoder_is_configured_);
+
+  // TODO: libiamf has an issue outputting 32 bit float samples, set to 16 bit
+  // for now.
+  int error = IAMF_decoder_set_bit_depth(decoder_, 16);
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR) << "IAMF_decoder_set_bit_depth() fails with error "
+                   << ErrorCodeToString(error);
+    return false;
+  }
+
+  error = IAMF_decoder_set_sampling_rate(decoder_, reader_.sample_rate());
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR) << "IAMF_decoder_set_sampling_rate() fails with error "
+                   << ErrorCodeToString(error);
+    return false;
+  }
+
+  if (kForceBinauralAudio || prefer_binarual_audio_) {
+    SB_DLOG(INFO) << "Configuring IamfAudioDecoder for binaural output";
+    error = IAMF_decoder_output_layout_set_binaural(decoder_);
+    if (error != IAMF_OK) {
+      SB_DLOG(ERROR)
+          << "IAMF_decoder_output_layout_set_binaural() fails with error "
+          << ErrorCodeToString(error);
+      return false;
+    }
+  } else {
+    // Default to stereo output. If kEnableSurroundAudio is true, set to a sound
+    // system matching, the platform's audio configuration, if available.
+    IAMF_SoundSystem sound_system = SOUND_SYSTEM_A;
+    if (kEnableSurroundAudio) {
+      SbMediaAudioConfiguration out_config;
+      SbMediaGetAudioConfiguration(0, &out_config);
+      int channels = std::max(out_config.number_of_channels, 2);
+      if (channels > 8 || channels < 1) {
+        SB_DLOG(ERROR) << "Can't create decoder with " << channels
+                       << " channels";
+        return false;
+      }
+      switch (channels) {
+        case 1:
+          sound_system = SOUND_SYSTEM_MONO;
+          break;
+        case 2:
+          // Stereo output.
+          sound_system = SOUND_SYSTEM_A;
+          SB_DLOG(INFO) << "Configuring IamfAudioDecoder for stereo output";
+          break;
+        case 6:
+          // 5.1 output.
+          sound_system = SOUND_SYSTEM_B;
+          SB_DLOG(INFO) << "Configuring IamfAudioDecoder for 5.1 output";
+          break;
+        case 8:
+          // 7.1 output.
+          sound_system = SOUND_SYSTEM_C;
+          SB_DLOG(INFO) << "Configuring IamfAudioDecoder for 7.1 output";
+          break;
+        default:
+          SB_NOTREACHED();
+      }
+    } else {
+      SB_DLOG(INFO) << "Defaulting to stereo output.";
+    }
+
+    error = IAMF_decoder_output_layout_set_sound_system(decoder_, sound_system);
```

Comment:
SbMediaGetAudioConfiguration() will return the audio output device capabilities.  I don't think we should set the output sound system based on the audio output devices directly. For example, if the input is stereo and we have 5.1 audio output system, we'll set output sound system to 5.1 for stereo input here. Is that expected? And what will happen if we do that?